numpy/core/src/multiarray/dtypemeta.c

```diff
@@ -698,6 +698,28 @@ object_common_dtype(
     return cls;
 }
 
+int
+object_fill_zero_value(PyArrayObject *arr)
+{
+    PyObject *zero = PyLong_FromLong(0);
+    PyArray_FillObjectArray(arr, zero);
```

Comment:
This should be `raw_array_assign_scalar` (or `PyArray_FillWithScalar`).  We should try to remove `PyArray_FillObjectArray`.